lib/operations/crud_write.js

```diff
@@ -0,0 +1,284 @@
+const defineAspects = operation.defineAspects;
+const OperationBase = operation.OperationBase;
+const applyRetryableWrites = require('../utils').applyRetryableWrites;
+const applyWriteConcern = require('../utils').applyWriteConcern;
+const handleCallback = require('../utils').handleCallback;
+const decorateWithCollation = require('../utils').decorateWithCollation;
+const toError = require('../utils').toError;
+
+const INSERT = 'insert';
+const UPDATE = 'update';
+const REMOVE = 'remove';
+
+const RETRYABLE_WIRE_VERSION = 6;
+
+function isSingleServer(topology) {
+  if (topology.type && topology.type === 'server') {
+    return true;
+  }
+
+  if (topology.description && topology.description.type === 'Single') {
+    return true;
+  }
+
+  return false;
+}
+
+function isRetryabilitySupported(topology) {
+  if (isSingleServer(topology)) {
+    return false;
+  }
+
+  const maxWireVersion = topology.lastIsMaster().maxWireVersion;
+  if (maxWireVersion < RETRYABLE_WIRE_VERSION) {
+    return false;
+  }
+
+  if (!topology.logicalSessionTimeoutMinutes) {
+    return false;
+  }
+
+  return true;
+}
+
+class CrudWriteOperation extends OperationBase {
+  constructor(model) {
+    super(model.options);
+    this.model = model;
+  }
+
+  canRetry(server, topology) {
+    return (
+      this.options.retryWrites &&
+      this.session &&
+      !this.session.inTransaction() &&
+      isRetryabilitySupported(topology) &&
+      this.model.canRetry()
+    );
+  }
+
+  enableRetry() {
+    this.model.options.willRetryWrite = true;
+    this.session.incrementTransactionNumber();
+  }
+
+  execute(server, callback) {
+    return this.model.execute(server, callback);
+  }
+}
+
+class CrudModel {
+  constructor(collection, options) {
+    options = applyRetryableWrites(options, collection.s.db);
+    options = applyWriteConcern(options, { db: collection.s.db, collection }, options);
+    options.serializeFunctions = options.serializeFunctions || collection.s.serializeFunctions;
+    if (collection.s.options.ignoreUndefined) {
+      options.ignoreUndefined = collection.s.options.ignoreUndefined;
+    }
+
+    this.options = options;
+    this.namespace = collection.s.namespace;
+  }
+
+  execute(server, callback) {
+    if (!this.type || !this.operations) {
+      // TODO: Throw something?
+    }
+    server[this.type](this.namespace.toString(), this.operations, this.options, (err, result) => {
+      if (callback == null) return;
+      if (err) return handleCallback(callback, err, null);
+      if (result == null) return handleCallback(callback, null, null);
+      if (result.result.code) return handleCallback(callback, toError(result.result));
+      if (result.result.writeErrors)
+        return handleCallback(callback, toError(result.result.writeErrors[0]));
+      // Return the results
+      handleCallback(callback, null, result);
+    });
+  }
+}
+
+defineAspects(CrudWriteOperation, [Aspect.EXECUTE_WITH_SELECTION, Aspect.RETRY]);
```

Comment:
given this approach, does `Aspect.RETRY` buy us anything? I don't think it's used?

Reply:
Yeah, I stopped using it. I think I'd eventually like to make `canRetry` and `applyRetry` on the `OperationBase`.